flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java

```diff
@@ -0,0 +1,234 @@
+		} catch (ExecutionException e) {
+			assertThat(e.getMessage(), containsString("Failed to convert JobResult to JobExecutionResult."));
+		}
+
+		assertThatMiniClusterIsShutdown();
+	}
+
+	@Test
+	public void testJobClientSavepoint() throws Exception {
+		JobGraph jobGraph = new JobGraph();
+		String savepointPath = temporaryFolder.getRoot().getAbsolutePath();
+
+		PerJobMiniCluster perJobMiniCluster = new PerJobMiniCluster(new Configuration(), config -> {
+			// Use a mock in this test case to test calling of the savepoint methods
+			miniCluster = Mockito.mock(MiniCluster.class);
+
+			CompletableFuture<JobSubmissionResult> jobSubmissionFuture = new CompletableFuture<>();
+			jobSubmissionFuture.complete(new JobSubmissionResult(jobGraph.getJobID()));
+			Mockito.when(miniCluster.submitJob(Mockito.any())).thenReturn(jobSubmissionFuture);
+
+			CompletableFuture<JobResult> jobResultFuture = new CompletableFuture<>();
+			jobResultFuture.complete(
+				new JobResult.Builder()
+					.jobId(jobGraph.getJobID())
+					.netRuntime(42L)
+					.build());
+			Mockito.when(miniCluster.requestJobResult(Mockito.any())).thenReturn(jobResultFuture);
+
+			return miniCluster;
+		});
+
+		JobClient jobClient = perJobMiniCluster.submitJob(jobGraph).get();
+
+		jobClient.triggerSavepoint(savepointPath);
+		Mockito.verify(miniCluster).triggerSavepoint(jobGraph.getJobID(), savepointPath, false);
+
+		jobClient.stopWithSavepoint(true, savepointPath);
+		Mockito.verify(miniCluster).stopWithSavepoint(jobGraph.getJobID(), savepointPath, true);
+	}
+
+	@Test
+	public void testSubmissionError() throws Exception {
+		PerJobMiniCluster perJobMiniCluster = initializeMiniCluster();
+
+		// JobGraph is not a valid job
+		JobGraph jobGraph = new JobGraph();
+
+		try {
+			perJobMiniCluster.submitJob(jobGraph).get();
+		} catch (ExecutionException e) {
+			// That's what we want
+			assertThat(e.getMessage(), containsString("Error submitting job to MiniCluster"));
+		}
+
+		assertThatMiniClusterIsShutdown();
+	}
+
+	@Test
+	public void testMultipleExecutions() throws Exception {
+		PerJobMiniCluster perJobMiniCluster = initializeMiniCluster();
+		{
+			JobClient jobClient = perJobMiniCluster.submitJob(getNoopJobGraph()).get();
+			jobClient.getJobExecutionResult(getClass().getClassLoader()).get();
+			assertThatMiniClusterIsShutdown();
+		}
+		{
+			JobClient jobClient = perJobMiniCluster.submitJob(getNoopJobGraph()).get();
+			jobClient.getJobExecutionResult(getClass().getClassLoader()).get();
+			assertThatMiniClusterIsShutdown();
+		}
+	}
+
+	@Test
+	public void testJobClientInteractionAfterShutdown() throws Exception {
+		PerJobMiniCluster perJobMiniCluster = initializeMiniCluster();
+		JobClient jobClient = perJobMiniCluster.submitJob(getNoopJobGraph()).get();
+		jobClient.getJobExecutionResult(getClass().getClassLoader()).get();
+		assertThatMiniClusterIsShutdown();
+
+		try {
+			jobClient.cancel();
```

Comment:
`fail` is missing.

Reply:
Thanks for catching these. I'm used to using `assertThrows` which is available in more recent versions of Junit. I've added a method similar to this.

Reply:
Yeah, we should bump our Junit dependency.